### ISSUE TITLE
Probabely a bug in MysqlParser.g4 where condition

### ISSUE THREAD

**Issue by yafngzh** [antlr/grammars-v4#1525]
![image](https://user-images.githubusercontent.com/5991357/66450124-ec74af80-ea89-11e9-9cbe-9230bfe981ec.png)

As shown in the image, when negative op and minus op are together, they are mistaken as minusminus.

**Comment by studentmain** [comment]
https://github.com/antlr/grammars-v4/blob/306eb4776ee3347a81caa4a6933033f5ad5e660a/mysql/Positive-Technologies/MySqlParser.g4#L2335-L2337

Is '--' really a mathOperator?

**Comment by mike-lischke** [comment]
This grammar has a number of flaws (beside being incomplete). Use the Oracle grammar instead for correct parsing. And the bug you see is not in the where condition, but expressions. `--` is not a valid operator in MySQL. Instead it is used for single line comments (when followed by a whitespace).

**Comment by yafngzh** [comment]
Thanks! @mike-lischke . I already used Oracle grammer and it's OK now. The [MysqlParser.g4](https://github.com/antlr/grammars-v4/blob/master/mysql/Positive-Technologies/MySqlParser.g4) is just a little misleading, maybe you can add the above explaination  in [README.md](https://github.com/antlr/grammars-v4/blob/master/mysql/Positive-Technologies/README.md) ?

**Comment by KvanTTT** [comment]
I renamed tag `mysql` to `mysql-PositiveTechnologies` and introduced `mysql-Oracle` tag to avoid confusion.

Also, I have an idea to port mysql grammar by @mike-lischke to other targets using a universal grammar approach that already used in [JavaScript](https://github.com/antlr/grammars-v4/tree/master/javascript), [Python](https://github.com/antlr/grammars-v4/tree/master/python), [PHP](https://github.com/antlr/grammars-v4/tree/master/php).

**Comment by mike-lischke** [comment]
@KvanTTT Your "universal grammar approach" isn't that universal actually, as you are relying on syntax that can be understood by any runtime (e.g. simple variable access, simple function calls). I mentioned this idea a while ago in the mailing list and the MySQL grammar only uses predicates which follow this idea. No need to introduce any special accessor methods, like your `p("get")`. It's not necessary that all grammars use the same methods in their actions (and usually also not possible due to different requirements). If all grammar authors take care to use a simple syntax then we can always easily support all languages that support them. If needed, we can also extend the simple macro processing ANTLR does (e.g for `$text`), which is better than introducing a new nomenclature.

**Comment by mike-lischke** [comment]
Right, I'm also thinking since a while already about how to avoid target language specific code in actions, but I have not found a good approach so far.

>   >If needed, we can also extend the simple macro processing ANTLR does (e.g for $text), which is better than introducing a new nomenclature.
> 
> Unfortunately, default simple macro does not help a lot. What do you mean by extending? Introduce external grammars preprocessing?

What I mean is to introduce more elements like `$text` that provide a language agnostic way to access values in the current context. ANTLR converts them to target language code via the ST4 templates that exist for each runtime language.

**Comment by KvanTTT** [comment]
> What I mean is to introduce more elements like $text that provide a language agnostic way to access values in the current context. ANTLR converts them to target language code via the ST4 templates that exist for each runtime language.

I think it's a good idea and I have similar thoughts, see https://github.com/antlr/antlr4/issues/1045 and https://github.com/KvanTTT/Articles/blob/master/Ideal-parser-generator/English/Abstract.md

**Comment by gmai2006** [comment]
Added fix to this # 1525 by removing the MINUSMINUS from the lexer.  Will open an issue for support mix cases and add a proposed grammar file.
https://github.com/gmai2006/grammar/
